README.md
# Ruby Zonefile - Parse and manipulate DNS Zone Files.

## Description

This class can read, manipulate and create DNS zone files. It supports A, AAAA, MX, NS, SOA, 
TXT, CNAME, PTR and SRV records. The data can be accessed by the instance method of the same
name. All except SOA return an array of hashes containing the named data. SOA directly returns the 
hash since there can only be one SOA information.

The following hash keys are returned per record type:

* SOA  
     :ttl, :primary, :email, :serial, :refresh, :retry, :expire, :minimumTTL
* A  
     :name, :ttl, :class, :host
* MX  
     :name, :ttl, :class, :pri, :host
* NS  
     :name, :ttl, :class, :host
* CNAME   
     :name, :ttl, :class, :host
* TXT  
     :name, :ttl, :class, :text
* A4 (AAAA)  
     :name, :ttl, :class, :host
* PTR  
     :name, :ttl, :class, :host
* SRV  
     :name, :ttl, :class, :pri, :weight, :port, :host
* DS  
     :name, :ttl, :class, :key_tag, :algorithm, :digest_type, :digest
* DNSKEY  
     :name, :ttl, :class, :flag, :protocol, :algorithm, :public_key
* RRSIG  
     :name, :ttl, :class, :type_covered, :algorithm, :labels, :original_ttl,
      :expiration, :inception, :key_tag, :signer, :signature
* NSEC  
     :name, :ttl, :class, :next, :types
* NSEC3  
     :name, :ttl, :class, :algorithm, :flags, :iterations, :salt, :next, :types
* NSEC3PARAM  
     :name, :ttl, :class, :algorithm, :flags, :iterations, :salt
* TLSA  
     :name, :ttl, :class, :certificate_usage, :selector, :matching_type, :data
* NAPTR  
     :name, :ttl, :class, :order, :preference, :flags, :service, :regexp, :replacement

# Examples

## Read a Zonefile

    zf = Zonefile.from_file('/path/to/zonefile.db')
  
    # Display MX-Records
    zf.mx.each do |mx_record|
      puts "Mail Exchagne with priority: #{mx_record[:pri]} --> #{mx_record[:host]}"
    end

    # Show SOA TTL
    puts "Record Time To Live: #{zf.soa[:ttl]}"
 
    # Show A-Records
    zf.a.each do |a_record|
      puts "#{a_record[:name]} --> #{a_record[:host]}"
    end

## Manipulate a Zonefile

    zf = Zonefile.from_file('/path/to/zonefile.db')
  
    # Change TTL and add an A-Record
  
    zf.soa[:ttl] = '123123'      # Change the SOA ttl
    zf.a << { :class => 'IN', :name => 'www', :host => '192.168.100.1', :ttl => 3600 }  # add A-Record
 
    # Setting PTR records (deleting existing ones)

    zf.ptr = [ { :class => 'IN', :name=>'1.100.168.192.in-addr.arpa', :host => 'my.host.com' },
               { :class => 'IN', :name=>'2.100.168.192.in-addr.arpa', :host => 'me.host.com' } ]

    # Increase Serial Number
    zf.new_serial

    # Print new zonefile
    puts "New Zonefile: \n#{zf.output}"

# Name attribute magic

Since 1.04 the :name attribute is preserved and returned as defined in a previous record if a zonefile entry
omits it. This should be the expected behavior for most users.
You can switch this off globally by calling Zonefile.preserve_name(false)


# Authors
 
Martin Boese, based on Simon Flack Perl library DNS::ZoneParse 

Andy Newton, patch to support various additional records

Julian Kornberger, patch to support TLSA records


CHANGELOG
1.00 - Initial Release
1.01 - Fixes
1.02 - Fixes
1.03 - Fixes TXT records, quotes are not treated anymore
1.05 - Adds support for TLSA records


tests/zonefile.rb
require 'test/unit'


$: << File.expand_path(File.dirname(__FILE__) + '/../lib')

require 'zonefile'

# Zonefile.preserve_name(false)

$zonefile = ARGV[0] || 'test-zone.db'

class TC_Zonefile  < Test::Unit::TestCase

  def setup
    @zf = Zonefile.from_file(File.dirname(__FILE__) + '/'+$zonefile, 'test-origin')
  end

  def swap # generate output and re-read @zf from it
    @zf = Zonefile.new(@zf.output, 'test-origin')
  end

  def test_empty
     zf = Zonefile.new
     zf.soa[:refresh] =  1234
     assert zf.empty?
  end
  
  def test_setter
    data = [ { :class => 'IN', :name => '123', :host => 'test' },
             { :name => '321', :hosts => 'test2' } ]
    @zf.ptr = data
    assert_equal 2, @zf.ptr.size    
    assert @zf.ptr[0][:host] == data[0][:host]
    assert @zf.ptr[1][:name] == data[1][:name]
    assert_raise(NoMethodError) do
        @zf.dont_exist(123,123,123)
    end
  end


  def test_soa
    assert_equal '86400', @zf.soa[:minimumTTL]
    assert_equal '691200', @zf.soa[:expire]
    assert_equal '3600', @zf.soa[:retry]
    assert_equal '10800', @zf.soa[:refresh]
    assert_equal '2000100501', @zf.soa[:serial]
    assert_equal 'support.dns-zoneparse-test.net.', @zf.soa[:email]
    assert_equal 'ns0.dns-zoneparse-test.net.', @zf.soa[:primary]
    
    begin
      @swap_soa = true
      swap
      test_soa
    end unless @swap_soa
  end

  def test_a
    assert_equal 9, @zf.a.size
    assert_equal 'mini', @zf.a.last[:name]
    assert_equal '10.0.0.7', @zf.a.last[:host]
    assert_equal '127.0.0.1', @zf.a.first[:host]

    a = @zf.a.find { |a| a[:host] == '10.0.0.3'}
    assert_equal '43200', a[:ttl]
    assert_equal 'www', a[:name].to_s   # name preserved

    begin
      @swap_a = true
      swap
      test_a
    end unless @swap_a
  end
  
  def test_preserve_name
    Zonefile.preserve_name(false)
    setup
    a = @zf.a.find { |a| a[:host] == '10.0.0.2'}
    assert_nil a[:name]             # no name preserved
    assert_nil @zf.nsec3[0][:name]  # same here
    Zonefile.preserve_name(true)
    setup
    a = @zf.a.find { |a| a[:host] == '10.0.0.2'}
    assert_equal 'www', a[:name]        # now name IS preserved
    assert_equal 'alfa.example.com.', @zf.nsec3[0][:name]  # same here
  end

  def test_mx
    assert_equal 2, @zf.mx.size
    assert_equal 10, @zf.mx.first[:pri]
    begin
      @swap_mx = true
      swap
      test_mx
    end unless @swap_mx
  end
  
  def test_cname 
    assert !!@zf.cname.find { |e| e[:host] == 'www' }
    begin
      @swap_cname = true
      swap
      test_cname
    end unless @swap_cname
  end
  
  def test_ns
    assert_equal 'ns0.dns-zoneparse-test.net.', @zf.ns[0][:host]
    assert_equal 'ns1.dns-zoneparse-test.net.', @zf.ns[1][:host]
    begin
      @swap_ns = true
      swap
      test_ns
    end unless @swap_ns
  end
  
  def test_txt
    puts @zf.txt.inspect
    assert_equal '"web;server"', @zf.txt[0][:text]
    assert_equal 'IN', @zf.txt[1][:class]
    assert_equal 'soup', @zf.txt[1][:name]
    assert_equal 'txta', @zf.txt[2][:name]
    assert_equal 'IN', @zf.txt[3][:class]
    assert_equal "\"t=y; o=-\"", @zf.txt[2][:text]
    assert_equal 'maxnet.ao', @zf.txt[3][:text]
    assert_equal '_kerberos', @zf.txt[3][:name]
    
    assert_equal 4, @zf.txt.size
    begin
      @swap_txt = true
      swap
      test_txt
    end unless @swap_txt
  end
  
  def test_a4
    assert_equal 'icarus', @zf.a4[0][:name]
    assert_equal 'IN', @zf.a4[0][:class]
    assert_equal 1, @zf.a4.size
    assert_equal 'fe80::0260:83ff:fe7c:3a2a', @zf.a4[0][:host]
    begin
      @swap_a4 = true
      swap
      test_a4
    end unless @swap_a4
  end
  
  def test_srv
    assert_equal '_sip._tcp.example.com.', @zf.srv[0][:name]
    assert_equal '86400', @zf.srv[0][:ttl]
    assert_equal '0', @zf.srv[0][:pri]
    assert_equal '5', @zf.srv[0][:weight]
    assert_equal '5060', @zf.srv[0][:port]
    assert_equal 'sipserver.example.com.', @zf.srv[0][:host]
    begin
      @swap_srv = true
      swap
      test_srv
    end unless @swap_srv
  end
  
  def test_serial_generator
    old = @zf.soa[:serial]
    new = @zf.new_serial
    assert new.to_i > old.to_i
    newer = @zf.new_serial
    assert newer.to_i - 1, new
    
    @zf.soa[:serial] = '9999889901'
    @zf.new_serial
    assert_equal '9999889902', @zf.soa[:serial]
  end

  def test_ptr
     assert_equal '12.23.21.23.in-addr.arpa', @zf.ptr[0][:name]
     assert_equal 'www.myhost.example.com.',  @zf.ptr[0][:host] 
     begin
       @swap_ptr = true
       swap
       test_ptr
     end unless @swap_ptr
  end

  def test_ds
    assert_equal "ds1", @zf.ds[0][:name]
    assert_equal 31528, @zf.ds[0][:key_tag]
    assert_equal "5", @zf.ds[0][:algorithm]
    assert_equal 1, @zf.ds[0][:digest_type]
    assert_equal "2274EACD70C5CD6862E1C0262E99D48D9FDEC271", @zf.ds[0][:digest]
    assert_equal "ds2", @zf.ds[1][:name]
    assert_equal 31528, @zf.ds[1][:key_tag]
    assert_equal "5", @zf.ds[1][:algorithm]
    assert_equal 1, @zf.ds[1][:digest_type]
    assert_equal "2BB183AF5F22588179A53B0A98631FAD1A292118", @zf.ds[1][:digest]
    begin
      @swap_ds = true
      swap
      test_ds
    end unless @swap_ds
  end

  def test_nsec
    assert_equal "alfa.example.com.", @zf.nsec[0][:name]
    assert_equal "host.example.com.", @zf.nsec[0][:next]
    assert_equal "A MX RRSIG NSEC TYPE1234", @zf.nsec[0][:types]
    begin
      @swap_nsec = true
      swap
      test_nsec
    end unless @swap_nsec
  end

  def test_nsec3
    assert_equal "1", @zf.nsec3[0][:algorithm]
    assert_equal "1", @zf.nsec3[0][:flags]
    assert_equal "12", @zf.nsec3[0][:iterations]
    assert_equal "aabbccdd", @zf.nsec3[0][:salt]
    assert_equal "2vptu5timamqttgl4luu7kg2leoaor3s", @zf.nsec3[0][:next]
    assert_equal "A RRSIG", @zf.nsec3[0][:types]
    begin
      @swap_nsec3 = true
      swap
      test_nsec3
    end unless @swap_nsec3
  end

  def test_nsec3param
    assert_equal "1", @zf.nsec3param[0][:algorithm]
    assert_equal "0", @zf.nsec3param[0][:flags]
    assert_equal "12", @zf.nsec3param[0][:iterations]
    assert_equal "aabbccdd", @zf.nsec3param[0][:salt]
    begin
      @swap_nsec3param = true
      swap
      test_nsec3param
    end unless @swap_nsec3param
  end

  def test_naptr
    assert_equal "urn.example.com.", @zf.naptr[0][:name]
    assert_equal 100, @zf.naptr[0][:order]
    assert_equal 50, @zf.naptr[0][:preference]
    assert_equal "\"s\"", @zf.naptr[0][:flags]
    assert_equal "\"http+N2L+N2C+N2R\"", @zf.naptr[0][:service]
    assert_equal "\"\"", @zf.naptr[0][:regexp]
    assert_equal "www.example.com.", @zf.naptr[0][:replacement]
    begin
      @swap_natpr = true
      swap
      test_naptr
    end unless @swap_natpr
  end

  def test_dnskey
    assert_equal "example.com.", @zf.dnskey[0][:name]
    assert_equal 256, @zf.dnskey[0][:flag]
    assert_equal 3, @zf.dnskey[0][:protocol]
    assert_equal "5", @zf.dnskey[0][:algorithm]
    pkey = <<PUBLIC_KEY.gsub( /\s+/,'').strip
           AQPSKmynfzW4kyBv015MUG2DeIQ3
           Cbl+BBZH4b/0PY1kxkmvHjcZc8no
           kfzj31GajIQKY+5CptLr3buXA10h
           WqTkF7H6RfoRqXQeogmMHfpftf6z
           Mv1LyBUgia7za6ZEzOJBOztyvhjL
           742iU/TpPSEDhm2SNKLijfUppn1U
           aNvv4w==
PUBLIC_KEY
    assert_equal pkey, @zf.dnskey[0][:public_key]
    assert_equal "example.net.", @zf.dnskey[1][:name]
    assert_equal 256, @zf.dnskey[1][:flag]
    assert_equal 3, @zf.dnskey[1][:protocol]
    assert_equal "5", @zf.dnskey[1][:algorithm]
    assert_equal pkey, @zf.dnskey[1][:public_key]
    begin
      @swap_dnskey = true
      swap
      test_dnskey
    end unless @swap_dnskey
  end

  def test_rrsig
    assert_equal "host.example.com.", @zf.rrsig[0][:name]
    assert_equal 'A', @zf.rrsig[0][:type_covered]
    assert_equal "5", @zf.rrsig[0][:algorithm]
    assert_equal 3, @zf.rrsig[0][:labels]
    assert_equal 86400, @zf.rrsig[0][:original_ttl]
    assert_equal 20030322173103, @zf.rrsig[0][:expiration]
    assert_equal 20030220173103, @zf.rrsig[0][:inception]
    assert_equal 2642, @zf.rrsig[0][:key_tag]
    assert_equal "example.com.", @zf.rrsig[0][:signer]
    sig = <<SIGNATURE.gsub( /\s+/,'').strip
        oJB1W6WNGv+ldvQ3WDG0MQkg5IEhjRip8WTr
        PYGv07h108dUKGMeDPKijVCHX3DDKdfb+v6o
        B9wfuh3DTJXUAfI/M0zmO/zz8bW0Rznl8O3t
        GNazPwQKkRN20XPXV6nwwfoXmJQbsLNrLfkG
        J5D6fwFm8nN+6pBzeDQfsS3Ap3o=
SIGNATURE
    assert_equal sig, @zf.rrsig[0][:signature]
    begin
      @swap_rrsig = true
      swap
      test_rrsig
    end unless @swap_rrsig
  end

  def test_tlsa
    assert_equal "_443._tcp.www.example.com.", @zf.tlsa[0][:name]
    assert_equal '86400', @zf.srv[0][:ttl]
    assert_equal 1, @zf.tlsa[0][:certificate_usage]
    assert_equal 1, @zf.tlsa[0][:selector]
    assert_equal 2, @zf.tlsa[0][:matching_type]

    sig = <<SIGNATURE.gsub( /\s+/,'').strip
        92003ba34942dc74152e2f2c408d29ec
        a5a520e7f2e06bb944f4dca346baf63c
        1b177615d466f6c4b71c216a50292bd5
        8c9ebdd2f74e38fe51ffd48c43326cbc
SIGNATURE
    assert_equal sig, @zf.tlsa[0][:data].gsub( /\s+/,'')
    
    begin 
      @swap_tlsa= true
      swap
      test_tlsa
    end unless @swap_tlsa
  end

  def test_origin
     assert_equal 'test-zone.db', @zf.origin
     swap
     assert_equal 'test-zone.db', @zf.origin     
  end
  
  
  def test_output
      puts @zf.data
      puts '==================='
      puts @zf.output  
  end

end


lib/zonefile/zonefile.rb
#
# = Ruby Zonefile - Parse and manipulate DNS Zone Files.
#
# == Description
# This class can read, manipulate and create DNS zone files. It supports A, AAAA, MX, NS, SOA, 
# TXT, CNAME, PTR and SRV records. The data can be accessed by the instance method of the same
# name. All except SOA return an array of hashes containing the named data. SOA directly returns the 
# hash since there can only be one SOA information.
#
# The following hash keys are returned per record type:
#
# * SOA
#    - :ttl, :primary, :email, :serial, :refresh, :retry, :expire, :minimumTTL
# * A
#    - :name, :ttl, :class, :host
# * MX
#    - :name, :ttl, :class, :pri, :host
# * NS
#    - :name, :ttl, :class, :host
# * CNAME
#    - :name, :ttl, :class, :host
# * TXT
#    - :name, :ttl, :class, :text
# * A4 (AAAA)
#    - :name, :ttl, :class, :host
# * PTR
#    - :name, :ttl, :class, :host
# * SRV
#    - :name, :ttl, :class, :pri, :weight, :port, :host
# * DS
#    - :name, :ttl, :class, :key_tag, :algorithm, :digest_type, :digest
# * DNSKEY
#    - :name, :ttl, :class, :flag, :protocol, :algorithm, :public_key
# * RRSIG
#    - :name, :ttl, :class, :type_covered, :algorithm, :labels, :original_ttl,
#      :expiration, :inception, :key_tag, :signer, :signature
# * NSEC
#    - :name, :ttl, :class, :next, :types
# * NSEC3
#    - :name, :ttl, :class, :algorithm, :flags, :iterations, :salt, :next, :types
# * NSEC3PARAM
#    - :name, :ttl, :class, :algorithm, :flags, :iterations, :salt
# * TLSA
#    - :name, :ttl, :class, :certificate_usage, :selector, :matching_type, :data
# * NAPTR
#    - :name, :ttl, :class, :order, :preference, :flags, :service, :regexp, :replacement
#
# == Examples
#
# === Read a Zonefile
#
#  zf = Zonefile.from_file('/path/to/zonefile.db')
#  
#  # Display MX-Records
#  zf.mx.each do |mx_record|
#     puts "Mail Exchagne with priority: #{mx_record[:pri]} --> #{mx_record[:host]}"
#  end
#
#  # Show SOA TTL
#  puts "Record Time To Live: #{zf.soa[:ttl]}"
#
#  # Show A-Records
#  zf.a.each do |a_record|
#     puts "#{a_record[:name]} --> #{a_record[:host]}"
#  end
#
#
# ==== Manipulate a Zonefile
#
#  zf = Zonefile.from_file('/path/to/zonefile.db')
#
#  # Change TTL and add an A-Record
#
#  zf.soa[:ttl] = '123123'	# Change the SOA ttl
#  zf.a << { :class => 'IN', :name => 'www', :host => '192.168.100.1', :ttl => 3600 }  # add A-Record
#
#  # Setting PTR records (deleting existing ones)
#
#  zf.ptr = [ { :class => 'IN', :name=>'1.100.168.192.in-addr.arpa', :host => 'my.host.com' },
#             { :class => 'IN', :name=>'2.100.168.192.in-addr.arpa', :host => 'me.host.com' } ]
#
#  # Increase Serial Number
#  zf.new_serial
#
#  # Print new zonefile
#  puts "New Zonefile: \n#{zf.output}"
#
# == Name attribute magic
#
# Since 1.04 the :name attribute is preserved and returned as defined in a previous record if a zonefile entry
# omits it. This should be the expected behavior for most users.
# You can switch this off globally by calling Zonefile.preserve_name(false)
#
# == Authors
# 
# Martin Boese, based on Simon Flack Perl library DNS::ZoneParse 
#
# Andy Newton, patch to support various additional records
#

class Zonefile

 RECORDS = %w{ mx a a4 ns cname txt ptr srv soa ds dnskey rrsig nsec nsec3 nsec3param tlsa naptr }
 attr :records
 attr :soa
 attr :data
# global $ORIGIN option
 attr :origin 
 # global $TTL option
 attr :ttl
 
 @@preserve_name = true

 # For compatibility: This can switches off copying of the :name from the 
 # previous record in a zonefile if found omitted. 
 # This was zonefile's behavior in <= 1.03 .  
 def self.preserve_name(do_preserve_name)
   @@preserve_name = do_preserve_name
 end
 
 def method_missing(m, *args)
   mname = m.to_s.sub("=","")
   return super unless RECORDS.include?(mname)
   
   if m.to_s[-1].chr == '=' then
     @records[mname.intern] = args.first
     @records[mname.intern]
   else 
     @records[m]
   end
 end


 # Compact a zonefile content - removes empty lines, comments, 
 # converts tabs into spaces etc...
 def self.simplify(zf)
    # concatenate everything split over multiple lines in parentheses - remove ;-comments in block
    zf = zf.gsub(/(\([^\)]*?\))/) { |m| m.split(/\n/).map { |l| l.gsub(/\;.*$/, '') }.join("\n").gsub(/[\r\n]/, '').gsub( /[\(\)]/, '') }

    zf.split(/\n/).map do |line|
        r = line.gsub(/\t/, ' ')
        r = r.gsub(/\s+/, ' ')
        		# FIXME: this is ugly and not accurate, couldn't find proper regex:
        		#   Don't strip ';' if it's quoted. Happens a lot in TXT records.
        (0..(r.length - 1)).find_all { |i| r[i].chr == ';' }.each do |comment_idx|
           if !r[(comment_idx+1)..-1].index(/['"]/) then
              r = r[0..(comment_idx-1)]
              break
           end
        end
        r
    end.delete_if { |line| line.empty? || line[0].chr == ';'}.join("\n")
 end


 # create a new zonefile object by passing the content of the zonefile
 def initialize(zonefile = '', file_name= nil, origin= nil)
   @data = zonefile
   @filename = file_name
   @origin = origin || (file_name ? file_name.split('/').last : '')
  
   @records = {}
   @soa = {}
   RECORDS.each { |r| @records[r.intern] = [] }
   parse
 end
 
 # True if no records (except sao) is defined in this file
 def empty?
   RECORDS.each do |r|
      return false unless @records[r.intern].empty?
   end
   true
 end
 
 # Create a new object by reading the content of a file
 def self.from_file(file_name, origin = nil)
    Zonefile.new(File.read(file_name), file_name.split('/').last, origin)
 end
 
 def add_record(type, data= {})
    if @@preserve_name then
      @lastname = data[:name] if data[:name].to_s != ''
      data[:name] = @lastname if data[:name].to_s == ''
    end
    @records[type.downcase.intern] << data
 end
 
 # Generates a new serial number in the format of YYYYMMDDII if possible
 def new_serial
   base = "%04d%02d%02d" % [Time.now.year, Time.now.month, Time.now.day ]

   if ((@soa[:serial].to_i / 100) > base.to_i) then
       ns = @soa[:serial].to_i + 1
       @soa[:serial] = ns.to_s
       return ns.to_s
   end
   
   ii = 0
   while (("#{base}%02d" % ii).to_i <= @soa[:serial].to_i) do
    ii += 1
   end
   @soa[:serial] = "#{base}%02d" % ii   
 end

 def parse_line(line)
    valid_name = /[\@a-z_\-\.0-9\*]+/i
    valid_ip6  = /[\@a-z_\-\.0-9\*:]+/i
    rr_class   = /\b(?:IN|HS|CH)\b/i
    rr_type    = /\b(?:NS|A|CNAME)\b/i
    rr_ttl     = /(?:\d+[wdhms]?)+/i
    ttl_cls    = Regexp.new("(?:(#{rr_ttl})\s)?(?:(#{rr_class})\s)?")
    base64     = /([\s\w\+\/]*=*)/i
    hexadeimal = /([\sA-F0-9]*)/i
    quoted     = /(\"[^\"]*\")/i

    data = {}
    if line =~ /^\$ORIGIN\s*(#{valid_name})/ix then
        @origin = $1
    elsif line =~ /^(#{valid_name})? \s*
                 #{ttl_cls}
                 (#{rr_type}) \s
                 (#{valid_name})
               /ix then
              (name, ttl, dclass, type, host) = [$1, $2, $3, $4, $5]
             add_record($4, :name => $1, :ttl => $2, :class => $3, :host => $5)
    elsif line=~/^(#{valid_name})? \s*
                #{ttl_cls}
                AAAA \s
                (#{valid_ip6})               
                /x then
              add_record('a4', :name => $1, :ttl => $2, :class => $3, :host => $4)
    elsif line=~/^(#{valid_name})? \s*
                 #{ttl_cls}
                 MX \s
                 (\d+) \s
                 (#{valid_name})
               /ix then
               add_record('mx', :name => $1, :ttl => $2, :class => $3, :pri => $4.to_i, :host => $5)
    elsif line=~/^(#{valid_name})? \s*
                 #{ttl_cls}
                 SRV \s
                 (\d+) \s
                 (\d+) \s
                 (\d+) \s
                 (#{valid_name})
               /ix
	       add_record('srv', :name => $1, :ttl => $2, :class => $3, :pri => $4, :weight => $5,
	                         :port => $6, :host => $7)
    elsif line=~/^(#{valid_name})? \s*
                #{ttl_cls}
                DS \s
                (\d+) \s
                (\w+) \s
                (\d+) \s
                #{hexadeimal}
                /ix
        add_record( 'ds', :name => $1, :ttl => $2, :class => $3, :key_tag => $4.to_i, :algorithm => $5,
                    :digest_type => $6.to_i, :digest => $7.gsub( /\s/,'') )
    elsif line=~/^(#{valid_name})? \s*
                #{ttl_cls}
                NSEC \s
                (#{valid_name}) \s
                ([\s\w]*)
                /ix
      add_record( 'nsec', :name => $1, :ttl => $2, :class => $3, :next => $4, :types => $5.strip )
    elsif line=~/^(#{valid_name})? \s*
                #{ttl_cls}
                NSEC3 \s
                (\d+) \s
                (\d+) \s
                (\d+) \s
                (-|[A-F0-9]*) \s
                ([A-Z2-7=]*) \s
                ([\s\w]*)
                /ix
      add_record( 'nsec3', :name => $1, :ttl => $2, :class => $3, :algorithm => $4, :flags => $5,
                   :iterations => $6, :salt => $7, :next => $8.strip, :types => $9.strip )
    elsif line=~/^(#{valid_name})? \s*
                #{ttl_cls}
                NSEC3PARAM \s
                (\d+) \s
                (\d+) \s
                (\d+) \s
                (-|[A-F0-9]*)
                /ix
      add_record( 'nsec3param', :name => $1, :ttl => $2, :class => $3, :algorithm => $4, :flags => $5,
                  :iterations => $6, :salt => $7 )
    elsif line=~/^(#{valid_name})? \s*
                #{ttl_cls}
                DNSKEY \s
                (\d+) \s
                (\d+) \s
                (\w+) \s
                #{base64}
                /ix
      add_record( 'dnskey', :name => $1, :ttl => $2, :class => $3, :flag => $4.to_i, :protocol => $5.to_i,
                  :algorithm => $6, :public_key => $7.gsub( /\s/,'') )
    elsif line=~/^(#{valid_name})? \s*
                #{ttl_cls}
                RRSIG \s
                (\w+) \s
                (\w+) \s
                (\d+) \s
                (\d+) \s
                (\d+) \s
                (\d+) \s
                (\d+) \s
                (#{valid_name}) \s
                #{base64}
                /ix
      add_record( 'rrsig', :name => $1, :ttl => $2, :class => $3, :type_covered => $4, :algorithm => $5,
                  :labels => $6.to_i, :original_ttl => $7.to_i, :expiration => $8.to_i, :inception => $9.to_i,
                  :key_tag => $10.to_i, :signer => $11, :signature => $12.gsub( /\s/,'')  )
    elsif line=~/^(#{valid_name}) \s*
                #{ttl_cls}
                TLSA \s
                (\d+) \s
                (\d+) \s
                (\d+) \s
                #{base64}
                /ix
      add_record( 'tlsa', :name => $1, :ttl => $2, :class => $3, :certificate_usage => $4.to_i,
                  :selector => $5.to_i, :matching_type => $6.to_i, :data => $7 )
    elsif line=~/^(#{valid_name})? \s*
                #{ttl_cls}
                NAPTR \s
                (\d+) \s
                (\d+) \s
                #{quoted} \s
                #{quoted} \s
                #{quoted} \s
                (#{valid_name})
                /ix
      add_record( 'naptr', :name => $1, :ttl => $2, :class => $3, :order => $4.to_i, :preference => $5.to_i,
                  :flags => $6, :service => $7, :regexp => $8, :replacement => $9 )
    elsif line=~/^(#{valid_name}) \s+
                 #{ttl_cls}
                 SOA \s+
                 (#{valid_name}) \s+
                 (#{valid_name}) \s*
                 \s*
                     (#{rr_ttl}) \s+
                     (#{rr_ttl}) \s+
                     (#{rr_ttl}) \s+
                     (#{rr_ttl}) \s+
                     (#{rr_ttl}) \s*
               /ix
            ttl = @soa[:ttl] || $2 || ''
            @soa[:origin] = $1
            @soa[:ttl] = ttl
            @soa[:primary] = $4
            @soa[:email] = $5
            @soa[:serial] = $6
            @soa[:refresh] = $7
            @soa[:retry] = $8
            @soa[:expire] = $9
            @soa[:minimumTTL] = $10

    elsif line=~ /^(#{valid_name})? \s*
                #{ttl_cls}
                PTR \s+
                (#{valid_name})
               /ix
            add_record('ptr', :name => $1, :class => $3, :ttl => $2, :host => $4)
    elsif line =~ /^(#{valid_name})? \s* #{ttl_cls} TXT \s+ (.*)$/ix
             add_record('txt', :name => $1, :ttl => $2, :class => $3, :text => $4.strip)
    elsif line =~ /\$TTL\s+(#{rr_ttl})/i 
            @ttl = $1
    end
 end

 def parse
    Zonefile.simplify(@data).each_line do |line|
        parse_line(line)      
    end
 end
 
 
 # Build a new nicely formatted Zonefile
 #
 def output
    out =<<-ENDH
;
;  Database file #{@filename || 'unknown'} for #{@origin || 'unknown'} zone.
;	Zone version: #{self.soa[:serial]}
;
#{self.soa[:origin]}		#{self.soa[:ttl]} IN  SOA  #{self.soa[:primary]} #{self.soa[:email]} (
				#{self.soa[:serial]}	; serial number
				#{self.soa[:refresh]}	; refresh
				#{self.soa[:retry]}	; retry
				#{self.soa[:expire]}	; expire
				#{self.soa[:minimumTTL]}	; minimum TTL
				)

#{@origin ? "$ORIGIN #{@origin}" : ''}
#{@ttl ? "$TTL #{@ttl}" : ''}
				
; Zone NS Records
ENDH
   self.ns.each do |ns|
     out <<  "#{ns[:name]}	#{ns[:ttl]}	#{ns[:class]}	NS	#{ns[:host]}\n"
   end
   out << "\n; Zone MX Records\n" unless self.mx.empty?
   self.mx.each do |mx|
     out << "#{mx[:name]}	#{mx[:ttl]}	#{mx[:class]}	MX	#{mx[:pri]} #{mx[:host]}\n"
   end
   
   out << "\n; Zone A Records\n" unless self.a.empty?
   self.a.each do |a|
        out <<  "#{a[:name]}	#{a[:ttl]}	#{a[:class]}	A	#{a[:host]}\n"
   end   

   out << "\n; Zone CNAME Records\n" unless self.cname.empty?
   self.cname.each do |cn|
     out << "#{cn[:name]}	#{cn[:ttl]}	#{cn[:class]}	CNAME	#{cn[:host]}\n"
   end  

   out << "\n; Zone AAAA Records\n" unless self.a4.empty?
   self.a4.each do |a4|
     out << "#{a4[:name]}	#{a4[:ttl]}	#{a4[:class]}	AAAA	#{a4[:host]}\n"
   end

   out << "\n; Zone TXT Records\n" unless self.txt.empty?
   self.txt.each do |tx|
     out << "#{tx[:name]}	#{tx[:ttl]}	#{tx[:class]}	TXT	#{tx[:text]}\n"
   end

   out << "\n; Zone SRV Records\n" unless self.srv.empty?
   self.srv.each do |srv|
     out << "#{srv[:name]}	#{srv[:ttl]}	#{srv[:class]}	SRV	#{srv[:pri]} #{srv[:weight]} #{srv[:port]}	#{srv[:host]}\n"
   end
   
   out << "\n; Zone PTR Records\n" unless self.ptr.empty?
   self.ptr.each do |ptr|
     out << "#{ptr[:name]}	#{ptr[:ttl]}	#{ptr[:class]}	PTR	#{ptr[:host]}\n"
   end

   out << "\n; Zone DS Records\n" unless self.ds.empty?
   self.ds.each do |ds|
     out << "#{ds[:name]} #{ds[:ttl]} #{ds[:class]} DS #{ds[:key_tag]} #{ds[:algorithm]} #{ds[:digest_type]} #{ds[:digest]}\n"
   end

   out << "\n; Zone NSEC Records\n" unless self.ds.empty?
   self.nsec.each do |nsec|
     out << "#{nsec[:name]} #{nsec[:ttl]} #{nsec[:class]} NSEC #{nsec[:next]} #{nsec[:types]}\n"
   end

   out << "\n; Zone NSEC3 Records\n" unless self.ds.empty?
   self.nsec3.each do |nsec3|
     out << "#{nsec3[:name]} #{nsec3[:ttl]} #{nsec3[:class]} NSEC3 #{nsec3[:algorithm]} #{nsec3[:flags]} #{nsec3[:iterations]} #{nsec3[:salt]} #{nsec3[:next]} #{nsec3[:types]}\n"
   end

   out << "\n; Zone NSEC3PARAM Records\n" unless self.ds.empty?
   self.nsec3param.each do |nsec3param|
     out << "#{nsec3param[:name]} #{nsec3param[:ttl]} #{nsec3param[:class]} NSEC3PARAM #{nsec3param[:algorithm]} #{nsec3param[:flags]} #{nsec3param[:iterations]} #{nsec3param[:salt]}\n"
   end

   out << "\n; Zone DNSKEY Records\n" unless self.ds.empty?
   self.dnskey.each do |dnskey|
     out << "#{dnskey[:name]} #{dnskey[:ttl]} #{dnskey[:class]} DNSKEY #{dnskey[:flag]} #{dnskey[:protocol]} #{dnskey[:algorithm]} #{dnskey[:public_key]}\n"
   end

   out << "\n; Zone RRSIG Records\n" unless self.ds.empty?
   self.rrsig.each do |rrsig|
     out << "#{rrsig[:name]} #{rrsig[:ttl]} #{rrsig[:class]} RRSIG #{rrsig[:type_covered]} #{rrsig[:algorithm]} #{rrsig[:labels]} #{rrsig[:original_ttl]} #{rrsig[:expiration]} #{rrsig[:inception]} #{rrsig[:key_tag]} #{rrsig[:signer]} #{rrsig[:signature]}\n"
   end

   out << "\n; Zone TLSA Records\n" unless self.tlsa.empty?
   self.tlsa.each do |tlsa|
     out << "#{tlsa[:name]} #{tlsa[:ttl]} #{tlsa[:class]} TLSA #{tlsa[:certificate_usage]} #{tlsa[:selector]} #{tlsa[:matching_type]} #{tlsa[:data]}\n"
   end

   out << "\n; Zone NAPTR Records\n" unless self.ds.empty?
   self.naptr.each do |naptr|
     out << "#{naptr[:name]} #{naptr[:ttl]} #{naptr[:class]} NAPTR #{naptr[:order]} #{naptr[:preference]} #{naptr[:flags]} #{naptr[:service]} #{naptr[:regexp]} #{naptr[:replacement]}\n"
   end

   out
 end

end

